Add format-selection decision framework page and case studies (#205)
cloudnativegeo/cloud-optimized-geospatial-formats-guide: decision-framework/case-studies.qmd

---
title: "Decision Framework Case Studies"
subtitle: "Six products run through the framework, five of them backed by measurements"
---

Each case study below runs a real product through the [decision framework](index.qmd). It shows where the product lands and why. Five of the six are backed by **measured benchmark results**.

> Read each case in two parts:

Reviewer: ```suggestion
Each case study includes:

1. Characteristics of the product
2. What the data is intended for
3. What path the case study follows through the decision tree
4. A size measurement for each candidate format
5. The decision outcome
```

> 1. **The path through the spine**: the control fork, the [topology question](index.qmd#topology), then intent, the object-size lever, maturity, and interpretation encoding.

Reviewer 2: why is topology question the only linked section?

> 2. **What the measurement showed.**

Reviewer 2: ```suggestion
2. **What the benchmark measurements showed.**
```

> The examples use public Copernicus and NASA missions. The [CNES Datalake](https://cnes.fr/en/projects/centre-de-calcul) appears as one worked example of a tiered object store. It is not the only kind of backend, and its tier numbers are not a general rule.

Reviewer 2: Did we plan to have a glossary, or is the decision framework the entirety. Just wondering as I had to think to recall what Tiered store meant in this case.

Reviewer 2: Maybe in the first paragraph we should just be more explicit that if a reader doesn't understand something to check the Decision Framework for an explanation.

Author: Key terms on the framework page covers vocabulary the framework uses in a specific way (control, topology, refinement, intent, the grouping lever, maturity, interpretation encoding). General cloud-native vocabulary belongs in the guide's [glossary](https://guide.cloudnativegeo.org/glossary.html), which is where the whole guide already looks.
I take both suggestions

> ::: {.callout-important}

Reviewer: ```suggestion
::: {.callout-note}
```

> ## The recurring lesson

Reviewer: ```suggestion
## Cloud-optimized != smaller storage size
```

> **Going cloud-optimized is rarely smaller than the source you started from.**

Reviewer: ```suggestion
**A cloud-optimized format is not necessarily smaller than a non-cloud-optimized version.**
```

> What it buys you is partial access, a spatial index, or a level of detail. Where two cloud-optimized candidates *do* differ in size, the cause is usually the **codec**, not the format.

Reviewer: ```suggestion
Cloud-optimized formats provide partial access via byte-range indices into the data, accessible via addressable metadata.
```

The second sentence doesn't follow -- this note/warning is describing the reason for a size difference between cloud-optimized and non-cloud-optimized, not comparing 2 cloud-optimized candidates.

> A second element explains much of the gap: **the extra bytes are the feature.** Overviews, levels of detail, and spatial indexes are structures the source file did not carry at all. They add volume, and they are exactly what buys fast display and cheap subsetting. The COG in case 1 is about 30 % larger than the delivered archive mostly because it carries overviews, and the row-group index in case 4 is the whole point of the conversion.

Reviewer: ```suggestion
**The extra bytes are the feature.** Overviews, levels of detail, and spatial indexes are structures the source file did not carry at all. They add volume, and they are exactly what buys fast display and cheap subsetting. The COG in case 1 is about 30% larger than the delivered archive mostly because it carries overviews, and the row-group index in case 4 is the whole point of the conversion.
```

:::

::: {.callout-note}
The numbers below are first measurements on a representative unit per mission, meaning one scene, one tile, or a few granules. They are illustrative, not a published benchmark. Two of the six cases have re-runs pending, and each one says so.

They were produced with [`cng-formats-benchmark`](https://github.com/developmentseed/cng-formats-benchmark), an open-source harness that measures object-size distribution, write, read, and display per format. You can describe your own datasets in its config files and reproduce this shape of comparison on your own data. See [Optional: make it quantitative](index.qmd#optional-make-it-quantitative).
:::

## 1. Per-scene optical raster: COG vs GeoZarr for visualization and distribution

**Product:** Sentinel-2 L2A. **Topology:** 2D raster per scene. **Source format:** geoTiff in a `.zip`. **Candidate set:** `{COG, GeoZarr}`.

**[Intent](index.qmd#intent).** Distribution and visualization. Both are served per scene, which is why this case does not need a cube.

> **The path.** Existing data, and the producer operates the processing chain, so this is a free choice. The topology opens `{COG, GeoZarr}` and the [intent](index.qmd#intent) decides between them. Per-scene **distribution and visualization** favors COG. **Time-series analysis** favors a GeoZarr cube. The [grouping lever](index.qmd#grouping-lever) is COG internal tiling, or a scene-size Zarr shard.

Reviewer 2: Thinking out loud, would this be easier to read as a bullet list?

1. Existing Data - Yes
2. Can you reprocess - Yes, the producer operates the processing chain, so this is a free choice.
3. What is the shape (Topology) - `{COG, GeoZarr}` 
4. [Intent](index.qmd#intent) - Distribution & Visualization favors COG, Time-series favors GeoZarr.
5. [Grouping](index.qmd#grouping-lever) - is COG internal tiling, or a scene-size Zarr shard.

Author: Yes thx, and it does something even better: the topology and the candidate set are currently stated twice per case, once in the metadata line and again in "The path". A trace list removes that.

Converting all six.

> **Measured, one scene.** Both layouts cleared the warm tier. COG came to 2625.7 MiB in 28 objects. A scene-size GeoZarr shard came to 1975.6 MiB in 26 objects, about 25 % smaller. That gap is the **codec**, Deflate against ZSTD, not the format. Against the 2023.7 MiB delivered `.zip`, the GeoZarr store is about source size (−2 %) and the COG is larger (+30 %), because the COG carries overviews.

Reviewer: What is the warm tier?

Author: clarified

**[Interpretation encoding](index.qmd#interpretation-encoding) is a real differentiator here.** The source reflectance is stored as scaled integers, and the nodata fill is left undeclared. COG puts the scale, the offset, and the nodata in side metadata that readers honor unevenly. Zarr v3 can carry all three as first-class array properties, using the `scale_offset` codec and a declared fill value, so every reader recovers the same physical value.

**Where it lands.** COG per scene for distribution and visualization.

> For temporal analysis GeoZarr is the right target, but **the arm measured here does not yet deliver it.** What was written is a scene-size store: one scene, so there is no time dimension to exploit. It proves the container and the encoding, because it comes out at roughly source size and it carries the scale, offset, and nodata correctly. A time-series query against it still has to open one store per date, which is the same fan-out the COG layout has.

Reviewer: I'm not sure what "the arm measured here" is referring to

Author: [Benchmarking](https://github.com/developmentseed/cng-formats-benchmark) jargon on my part. fixing

**The path.** Existing data, reprocessable, so a free choice. A year of RTC over a region is naturally a `time × y × x × pol` cube. On a Zarr cube, a time-series slice or an area-of-interest subset is one range-addressable query. Across thousands of per-timestamp COGs, the same request becomes a large fan-out. Sharding reconciles fine-grained chunk access with the backend's preference for large objects.

**Measured, one tile and confirmed at area scale.** COG came to 728 MiB per tile and a scene-size GeoZarr shard to 545 MiB. Both cleared the cold tier. GeoZarr is again about 25 % smaller, and again this is the codec: Deflate at 3.70x against ZSTD at 4.98x. The result held over a 16-tile block, at 16498 MiB for COG and 12344 MiB for GeoZarr. Against the 496 MiB delivered `.zip` both are larger, COG by 47 % and GeoZarr by 10 %. Cloud-optimized costs storage here, and buys the range access that the opaque zip cannot give at all.

::: {.callout-warning}
## A benchmark caveat, not a format verdict

> The GeoZarr store read and displayed much slower than the COG, but it was written **with no multiscale overview pyramid**. A store without multiscales is not a faithful GeoZarr, and the missing pyramid is the reason a tile server has to downsample full-resolution chunks on request. The slow read is also in large part a reader effect, from the library used rather than the format.

Reviewer: >A store without multiscales is not a faithful GeoZarr

Is this true? I thought multiscales were optional?

Author: relicate of the inital buggy becnhmark. Fixing here

> Both issues are being fixed for the final campaign, so that display is compared on equal terms. Treat the timing numbers from this arm as provisional.

Reviewer 2: I wonder if and when we should mention the example of the Xarray defaults being changed to better leverage cloud optimized formats. https://github.com/pydata/xarray/pull/11216 maybe not relevant here since this was for h5netcdf driver ...

Author: Checked and thanks for the heads up. It does have a precise home in the "Not yet possible" section, which says HDF5 became viable through a change in write strategy and reader tooling, and #11216 is exactly that second half, on exactly that format. linking it there.

:::

## 3. Point cloud: netCDF to COPC

**Product:** SWOT L2_HR Pixel Cloud. **Topology:** point cloud. **Source format:** netCDF. **Candidate set:** `{COPC, GeoParquet}`.

**[Intent](index.qmd#intent).** Distribution, visualization, and data fusion. The fusion intent is the one that matters most here: these points get combined with other water products, so a reader has to recover the same physical value per point and subset the same area cheaply. That is what makes the missing spatial index in the source a real cost rather than an inconvenience.

**The path.** Existing data, reprocessable, so a free choice. The point-cloud topology gives COPC, which is a clustered octree inside a LAZ file. The netCDF source has no spatial index and no level of detail, so reading a small geographic subset means scanning the whole granule.

**Measured, content-complete, three granules.** A COPC carrying all per-point variables was about 16 % **smaller** than the source netCDF. This is the one clear size win among the cases, and it comes from the codec: LASzip at 3.72x against netCDF Deflate at 2.56x. Each granule cleared the tier on its own.

The real gain is what the netCDF lacks:

> - A **spatial index**. A bounding-box query reads only the overlapping octree nodes.

Reviewer 2: Not sure if it matters in this example but COPC has a 3D spatial index(x,y,z), so you can also filter by z (height). I'm not sure Geoparquet can do that easily yet.

Author: Good catch, especially for this use case. water surface elevation is the measurement for PIXC so an elevation-range query is a real access pattern. Updating the bullet to say the octree indexes x, y and z.
One correction on GeoParquet: the 1.1 spec does allow it  so it's possible, just optional.

- A **level of detail**. Octree levels 0 to 11 are the point-cloud analogue of COG overviews.

::: {.callout-caution}
An earlier measurement on this product looked like a 97 % size saving. It had dropped most of the per-point variables. Size comparisons must be like for like on content.
:::

**Where it lands.** COPC. The cloud-optimized value is partial access plus level of detail, with a modest size win on top. GeoParquet is the alternative where columnar analytics dominate.

## 4. Vector and tabular: shapefile to GeoParquet

**Product:** SWOT L2_HR Lake Single Pass. **Topology:** vector and tabular. **Source format:** shapefile. **Candidate set:** `{GeoParquet, FlatGeobuf}`.

> **[Intent](index.qmd#intent).** Distribution, download, and data fusion. Note that visualization is *absent* from this mix, and download is present. That combination is why the answer is not automatically the columnar option: download rewards a single streamable file, which is FlatGeobuf's strength, while fusion rewards cheap joins on a fixed schema, which is GeoParquet's. The fusion weight is what settles it.

Reviewer 2: FlatGeobuf is also better for spatial subsetting if you want whole rows (all columns), not just because it's streamable. In fact streamable favors visualization more the download. Also since FlatGeobuf does not have compression, I think Geoparquet actually is the better Download (whole file). Scale of the file is pretty important too - once you go over a certain size flatgeobuf is somewhat unusable where geoparquet can be sharded into several files, which doesn't matter here because the files at tiny.

Author: Thx. The initial reasoning had download and streaming backwards. Download favors GeoParquet, since FlatGeobuf can't compress without breaking its index. Streaming is a visualization property, progressive rendering, and visualization is absent from this mix. 

One clarification: the +14 % is measured against the source shapefile, not against FlatGeobuf. GeoParquet is the smaller of the two, so your compression point and that number aren't in conflict. The measured line now states both totals on the same basis (35.6 vs 31.1 MiB for two passes)

Net: GeoParquet still wins, on download and fusion, so less finely balanced than I'd made it sound.

**The path.** Existing data, reprocessable, so a free choice. A 100 m raster granule covers a small tile, about 160 km across.

**Measured, single variable, so not yet like for like.** On the primary water-surface-elevation variable, COG came to 4.3 MiB per granule and a GeoZarr shard to 0.8 MiB on average, because sharding fragments at this size. Both sit far below the warm-tier floor, so the lever is **aggregation**, and at this small size **COG is the cleaner carrier**: one tiled object that also holds overviews. An all-variable re-run is needed before quoting a size, for the same reason as case 3.

The natural aggregation unit is a spatial mosaic, per grid zone and per orbit cycle. That turns a wall of tiny granules into one range-addressable object.

**Where it lands.** COG for the small raster, plus collection-level aggregation. GeoZarr is the alternative to compare once the pyramid and all variables are in place.

This case is the counterpart to cases 1 and 2. There, one large band already clears the tier, so you group *within* a product. Here the products are small, so you group *above* a single product.

## 6. Existing archive to a virtual store

> **Topology:** any. **Candidate set:** a virtual Zarr store.

Reviewer 2: I don't think this is "any" at least not without clarifying what formats are archival. You can't Virtual Zarr many vector formats.

Also we should be clear on the "can't reprocess" usually applies to very large datasets, that duplicative storage and reprocessing times would be prohibitive. This is also rarely the case for many vector datasets.

Author: Topology: any is gone in previous edits.
Your second point is useful one: "cannot reprocess" is normally about scale, not permission. That's exactly why it's an array story and rarely a vector one. Added as a short paragraph, which also makes the branch predictable in advance instead of something you discover at the end.

This case has no benchmark. It illustrates the "existing data, cannot reprocess" branch of the control fork.

An immutable archive, for example HDF5 or netCDF that you cannot reprocess, can be surfaced as a virtual Zarr store with [VirtualiZarr, Kerchunk, or Icechunk](../zarr/virtual-zarr.qmd). This gives range-based access without rewriting the data.

The result is more limited than a native store. Chunk sizes are whatever the source chose, you cannot add overviews, and you inherit the source's codec. It is still much better than whole-file download, and it is often the only option available.

> This case is also where the ["unlock this"](index.qmd#dead-ends-are-unlock-this) framing matters. If the source cannot even be virtualized, name what is missing rather than declaring the case impossible.

Reviewer 2: we might need to make reference to some more examples of data with "quirks" that are not currently virtualizable.

Author: Agreed. Right now, I do not have a real world use case to document. That's something a new use case owner could add later.
